hudi-spark/src/main/scala/org/apache/hudi/HudiMergeOnReadRDD.scala

```diff
@@ -0,0 +1,195 @@
+    }
+  }
+
+  override protected def getPartitions: Array[Partition] = {
+    hudiRealtimeFileSplits.zipWithIndex.map(file => HudiMergeOnReadPartition(file._2, file._1)).toArray
+  }
+
+  private def getConfig(): Configuration = {
+    broadcastedConf.value.config
+  }
+
+  private def read(partitionedFile: PartitionedFile,
+                   readFileFunction: PartitionedFile => Iterator[Any]): Iterator[InternalRow] = {
+    val fileIterator = readFileFunction(partitionedFile)
+    val rows = fileIterator.flatMap(_ match {
+      case r: InternalRow => Seq(r)
+      case b: ColumnarBatch => b.rowIterator().asScala
+    })
+    rows
+  }
+
+  private def unMergeFileIterator(split: HudiMergeOnReadFileSplit,
+                                  readFileFunction: PartitionedFile => Iterator[Any]): Iterator[InternalRow] =
+    new Iterator[InternalRow] {
+      private val dataFileIterator = read(split.dataFile, readFileFunction)
+      private val logSchema = getLogAvroSchema(split)
+      private val sparkTypes = SchemaConverters.toSqlType(logSchema).dataType.asInstanceOf[StructType]
+      private val converter = new AvroDeserializer(logSchema, sparkTypes)
+      private val hudiLogRecords = scanLog(split, logSchema).getRecords
+      private var hudiLogRecordsIterator = hudiLogRecords.keySet().iterator().asScala
+
+      override def hasNext: Boolean = {
+        dataFileIterator.hasNext || hudiLogRecordsIterator.hasNext
+      }
+
+      override def next(): InternalRow = {
+        if (dataFileIterator.hasNext) {
+          dataFileIterator.next()
+        } else {
+          val curAvrokey = hudiLogRecordsIterator.next()
+          val curAvroRecord = hudiLogRecords.get(curAvrokey).getData.getInsertValue(logSchema).get()
+          converter.deserialize(curAvroRecord).asInstanceOf[InternalRow]
+        }
+      }
+    }
+
+  private def mergeFileIterator(split: HudiMergeOnReadFileSplit,
+                                readFileFunction: PartitionedFile => Iterator[Any]): Iterator[InternalRow] =
+    new Iterator[InternalRow] {
+      private val dataFileIterator = read(split.dataFile, readFileFunction)
+      private val logSchema = getLogAvroSchema(split)
+      private val sparkTypes = SchemaConverters.toSqlType(logSchema).dataType.asInstanceOf[StructType]
+      private val converter = new AvroDeserializer(logSchema, sparkTypes)
+      private val hudiLogRecords = scanLog(split, logSchema).getRecords
+      private var parquetFinished = false
+      private var logRecordToRead = hudiLogRecords.keySet()
+      private var hudiLogRecordsIterator: Iterator[String] = _
+
+      override def hasNext: Boolean = {
+        if (dataFileIterator.hasNext) {
+          true
+        } else {
+          if (!parquetFinished) {
+            parquetFinished = true
+            hudiLogRecordsIterator = logRecordToRead.iterator().asScala
+          }
+          hudiLogRecordsIterator.hasNext
```

Comment:
this seems to indicate that we will keep scanning the remaining entries in the log and hand them out if `dataFileIterator` runs out. We need to be careful about how it interplays with split generation. specifically, only works if the each base file has only 1 split.. 
```
HudiMergeOnReadFileSplit(partitionedFile, logPaths, latestCommit,
         metaClient.getBasePath, maxCompactionMemoryInBytes, skipMerge)
```

here `partitionedFile` has to be a single file and not an input Split. otherwise we will face an issue that the log entry belongs to a different split and the ultimate query will have duplicates.

Reply:
Yes, this `HudiMergeOnReadFileSplit` is based on we pack one `baseFile` into one `partitionedFile` during the `buildFileIndex`. I believe Spark won't split this secretly somewhere...

Reply:
FileFormat could break up a file into splits I think. We can wait for @umehrot2 also to chime in here.

Reply:
One partitioned file will go into one split if I understand this correctly https://github.com/apache/spark/blob/branch-2.4/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala#L355
That would be difficult to handle if Spark split one parquet into two splits somewhere else.

Reply:
@vinothchandar @garyli1019 Spark's FileFormat uses can split the parquet files up if there are multiple row groups in the file. In Spark's implementation one `PartitionedFile` is not a complete file, that is why you can see the start position and length in that.

But in both this implementation and in bootstrap case we are packaging complete file as a split, to avoid the complexity of dealing with partial splits.